Set up CI with Azure Pipelines

jcampbell opened the pull request.

jcampbell committed:
[MAINTENANCE] Fix missing library tests

Signed-off-by: James Campbell <james.p.campbell@gmail.com>

jcampbell committed:
[MAINTENANCE] Improve contributing documentation for linting

Signed-off-by: James Campbell <james.p.campbell@gmail.com>

jcampbell committed:
[MAINTENANCE] Clarify test dependencies

Signed-off-by: James Campbell <james.p.campbell@gmail.com>

abegong commented on great_expectations/datasource/batch_kwargs_generator/batch_kwargs_generator.py:
```diff
@@ -9,7 +9,7 @@
 
 class BatchKwargsGenerator(object):
-    """
+    r"""
```
What does this do?

jcampbell replied:
It makes all your problems go away.

Try it.

Just once can't hurt.

(But the serious answer is that it makes python not treat `\` as an escape character, which is nice when you want to write a regex, because then you don't escape your escape. That doesn't *really* matter in this case since we don't operate on the docstring in general, but we *do* sometimes parse our own docstrings now for feature annotations, and it gets rid of warnings in the case of illegal would-be escaping.

abegong commented on docs/contributing/style_guide.rst:
```diff
@@ -24,7 +24,22 @@ code
 * **Experimental classes should log an experimental warning when initialized**: "Warning: great_expectations.some_module.SomeClass is experimental. Methods, APIs, and core behavior may change in the future."
 * **Docstrings are highly recommended**. We use the Sphinx's `Napoleon extension <http://www.sphinx-doc.org/en/master/ext/napoleon.html>`__ to build documentation from Google-style docstrings.
-* **Lint your code**. Our CI system will check using ``black`` and ``isort``. We have a git pre-commit configuration in the repo, so you can just run ``pre-commit install`` to automatically run your changes through the linting process before submitting.
+* **Lint your code**. Our CI system will check using ``black`` and ``isort``.
```
I'm not clear on how linting currently works. I haven't been linting, and that doesn't seem to have raised any issues.

jcampbell replied:
You probably have been either perfectly writing well-formatted code or are using an IDE that lints magically for you?

abegong replied:
Either way, I appear to be magical.

abegong commented on azure-pipelines.yml:
```diff
@@ -0,0 +1,169 @@
+              python.version: '3.7'
+              pandas.version: 'latest'
+
+        # TODO: refactor steps into a task
```
😱

jcampbell replied:
hahahahahaah !!! OH NO!! I HAVE BEEN HAD!

abegong commented on docs/contributing/style_guide.rst:
```diff
@@ -24,7 +24,22 @@ code
+        pip install -r requirements-dev.txt
+        pre-commit install --install-hooks
+
+  - To manually run the linter in your environment, for example if you have already committed files that are rejected by the linting test, you can run tests manually:
```
Non-blocking: I read this sentence three times and I still don't understand it.

jcampbell replied:
```suggestion
  - If you have already committed files but are seeing errors during the continuous integration tests, you can run tests manually:
```

jcampbell replied:
Please commit above if it addresses the question...

abegong commented on great_expectations/cli/datasource.py:
```diff
@@ -583,7 +583,7 @@ def _collect_postgres_credentials(default_credentials=None):
         default_credentials = {}
 
-    credentials = {"drivername": "postgres"}
+    credentials = {"drivername": "postgresql"}
```
yikes.

jcampbell replied:
FWIW, it's not awful -- postgres still works for now but is deprecated.

abegong commented on great_expectations/datasource/batch_kwargs_generator/batch_kwargs_generator.py:
```diff
@@ -202,7 +201,7 @@ def _build_batch_kwargs(self, batch_parameters):
 
     # TODO: deprecate generator_asset argument
-    def yield_batch_kwargs(self, generator_asset=None, data_asset_name=None, **kwargs):
+    def yield_batch_kwargs(self, data_asset_name=None, generator_asset=None, **kwargs):
```
Isn't this a breaking change?

`yield_batch_kwargs("foo", "bar")`
`yield_batch_kwargs("bar", "foo")`

jcampbell replied:
In this case, no, because one cannot provide both generator_asset and data_asset_name (already)...essentially you have to be using kwargs for anything beyond the first parameter, and so this just allows you to use the first parameter without getting a (needless) deprecation warning.

abegong commented on requirements-dev.txt:
```diff
@@ -1,35 +1,37 @@
-termcolor>=1.1.0
-tzlocal>=1.2
-xlrd>=1.1.0
+altair>=4.0.0,<5  # package
```
```suggestion
# A lot of Great Expectations' codebase and tests are tied to specific databases, cloud providers, etc.
# In this file, comments on each row indicate which tests that package is used for.
# To run tests for smaller subsets of infrastructure, please look at other requirements-dev-*.txt files.

altair>=4.0.0,<5  # package
```

I'd suggest adding a comment to explain the comments.

abegong reviewed (changes requested):
This looks awesome.

I've flagged one breaking API change in `yield_batch_kwargs`. It doesn't look like it affects anything in this PR, so I suggest rolling it back, then merging.

jcampbell committed:
# This is a combination of 2 commits.
# This is the 1st commit message:

[MAINTENANCE] Add support for azure pipelines

Signed-off-by: James Campbell <james.p.campbell@gmail.com>

# The commit message #2 will be skipped:

# Update pipeline container

jcampbell committed:
[MAINTENANCE] Add support for azure pipelines

Signed-off-by: James Campbell <james.p.campbell@gmail.com>

jcampbell committed:
Update requirements-dev.txt

Co-authored-by: Abe Gong <abegong@users.noreply.github.com>

abegong committed:
Update docs/contributing/style_guide.rst

Co-authored-by: James Campbell <james.p.campbell@gmail.com>

abegong reviewed (approved):
Are you **sure** you want to make the yield_batch_kwargs change? I get why it's probably non-breaking, but still...

Other than that, LGTM.

jcampbell committed:
Remove extra warning catch attempt

Signed-off-by: James Campbell <james.p.campbell@gmail.com>

jcampbell committed:
Merge remote-tracking branch 'origin/develop' into azure-pipelines

jcampbell committed:
Merge branch 'develop' into azure-pipelines
